### PR TITLE
Debug ekf

### PR DESCRIPTION
**Pull request by khancyr**
This allow to disable the pragma optimisation in EKF code

### CODE REVIEW

**Comment by peterbarker** [comment]
Any reason not to base this off the existing --debug?

**Comment by tridge** [comment]
I agree with Peter, I think --debug covers this unless there is some reason not to use it

**Comment by khancyr** [comment]
The only reason I saw, was that is wasn't done before so I add an extra command, but it would be better to be on the DEBUG one ! I will change that

**Comment by tridge** [comment]
why not the existing --debug configure option?

**Comment by khancyr** [comment]
I have revert to use the --debug parameter but you mean using the DEBUG macro and the one I created DEBUG_EKF_NO_O2  ?

File: Tools/ardupilotwaf/boards.py [hunk 1]
```diff
@@ -154,6 +154,9 @@ def configure_env(self, cfg, env):
                 '-g',
                 '-O0',
             ]
+            env.DEFINES.update(
+                DEBUG_EKF_NO_O2 = 1,
```
Comment (tridge) [inline Tools/ardupilotwaf/boards.py:158]: I'd rather a HAL_DEBUG_BUILD define

File: libraries/AP_Math/definitions.h [hunk 2]
```diff
@@ -27,6 +27,11 @@
   #define MATH_CHECK_INDEXES 0
 #endif
 
+// Disable #pragma GCC optimize("O2") on EKF2/3
```
Comment (tridge) [inline libraries/AP_Math/definitions.h:30]: don't need this

File: libraries/AP_NavEKF2/AP_NavEKF2_core.h [hunk 3]
```diff
@@ -20,11 +20,13 @@
  */
 #pragma once
 
+#if CONFIG_HAL_BOARD == HAL_BOARD_SITL && DEBUG_EKF_NO_O2==1
```
Comment (tridge) [inline libraries/AP_NavEKF2/AP_NavEKF2_core.h:23]: just #ifndef HAL_DEBUG_BUILD

File: libraries/AP_NavEKF3/AP_NavEKF3_core.h [hunk 4]
```diff
@@ -19,7 +19,11 @@
  */
 #pragma once
 
+
+#if CONFIG_HAL_BOARD == HAL_BOARD_SITL && DEBUG_EKF_NO_O2==1
```
Comment (tridge) [inline libraries/AP_NavEKF3/AP_NavEKF3_core.h:23]: #ifndef HAL_DEBUG_BUILD

**Comment by khancyr** [comment]
good to go !

File: libraries/AP_NavEKF3/AP_NavEKF3_core.h [hunk 5]
```diff
@@ -19,7 +19,10 @@
  */
 #pragma once
 
-#pragma GCC optimize("O2")
+
+#ifndef HAL_DEBUG_BUILD
```
Comment (peterbarker) [inline libraries/AP_NavEKF3/AP_NavEKF3_core.h:23]: This really ought to be `#if !`, not `#ifndef`

Really annoying when you change a 1 to a 0 and it has no effect.

Comment (khancyr) [reply libraries/AP_NavEKF3/AP_NavEKF3_core.h]: done

**Comment by khancyr** [comment]
master : 
`Tools/autotest/autotest.py build.Copter test.Copter  2764,48s user 659,04s system 120% cpu 47:16,17 total`

PR:
`Tools/autotest/autotest.py build.Copter test.Copter  2800,45s user 657,32s system 122% cpu 47:10,61 total`

so that is fine

**Comment by peterbarker** [comment]
Merged, thanks!

**Pull request merged**